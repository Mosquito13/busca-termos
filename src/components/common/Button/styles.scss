@import '../../../sass/styles.scss';

.button {
  background-color: transparent;
  border-radius: $radius;
  font-size: inherit;
  font-family: inherit;
  white-space: nowrap;
  padding: 1rem 1.5rem;
  display: flex;
  justify-content: center;
  align-items: center;
  cursor: pointer;
  outline: none;
  transition: all 0.2s;
  height: 4rem;

  @include themify() {
    border: 1px solid themed('color-text');
    color: themed('color-text');
  }

  &:focus {
    @include themify() {
      box-shadow: 0px 0px 5px themed('color-text');
    }
  }

  &:hover {
    @include themify() {
      color: themed('color-base');
      background-color: themed('color-text');
      box-shadow: 0px 0px 5px themed('color-text');
    }
  }

  &:active {
    @include themify() {
      color: themed('color-base');
      background-color: themed('color-text');
      box-shadow: 0px 0px 3px themed('color-text');
    }
    transform: translateY(1px);
  }

  &--borderless {
    border-color: transparent !important;
    box-shadow: none;
  }

  &--with-icon,
  &--loading {
    padding: 0.5rem 1rem;
    line-height: 1rem;
  }

  &--disabled {
    box-shadow: none;
    cursor: not-allowed;

    @include themify() {
      color: themed('color-text-disabled');
      border-color: themed('color-text-disabled');
    }

    &:hover,
    &:active {
      @include themify() {
        color: themed('color-text-disabled');
        background-color: transparent;
        box-shadow: none;
        transform: none;
      }
    }
  }

  &__content {
    transition: transform 0.2s;
  }

  &:active > &__content {
    transform: translateY(1px);
  }

  &--disabled:active > &__content {
    transform: none !important;
  }
}
