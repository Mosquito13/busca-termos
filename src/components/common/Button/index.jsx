import { useCallback } from 'react';
import { useSelector } from 'react-redux';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { ClipLoader } from 'react-spinners';

import Icon, { RESPONSIVE } from '../Icon';
import settingsSelectors from '../../../selectors/settings';

import './styles.scss';

const getContent = (text, icon, loading) => {
  if (loading) {
    return <ClipLoader size={25} color="currentColor" />;
  }

  return icon || text;
};

const Button = ({ text, icon, marker, tooltip, onClick, loading, disabled, borderless }) => {
  const compactLayout = useSelector(settingsSelectors.isCompactLayout);

  const classes = classNames(
    'button',
    borderless && 'button--borderless',
    icon && 'button--with-icon',
    loading && 'button--loading',
    disabled && 'button--disabled'
  );

  const handleClick = useCallback(() => {
    if (loading) {
      return false;
    }

    onClick();
  }, [loading, onClick]);

  return (
    <button
      className={classes}
      title={tooltip}
      onClick={handleClick}
      disabled={disabled}
    >
      <div className="button__content">{getContent(text, icon, loading)}</div>
      {marker && (
        <div className="button__marker">
          <Icon size={RESPONSIVE}>
            {!compactLayout && marker}
          </Icon>
        </div>
      )}
    </button>
  );
};

Button.propTypes = {
  /**
   * Texto exibido dentro do botao
   */
  text: PropTypes.string,
  /**
   * Icone exibido dentro do botao
   */
  icon: PropTypes.node,
  /**
   * Hint do botao
   */
  tooltip: PropTypes.string,
  /**
   * Acao ao clicar
   */
  onClick: PropTypes.func,
  /**
   * Indica se esta carregando, nao dispara o onClick se estiver carregando
   */
  loading: PropTypes.bool,
  /**
   * Indica se esta desabilitado
   */
  disabled: PropTypes.bool,
  /**
   * Indica se o botao deve ter borda invisivel
   */
  borderless: PropTypes.bool
};

export default Button;
