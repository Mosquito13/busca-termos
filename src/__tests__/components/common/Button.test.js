import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import * as reactRedux from 'react-redux';
import { FiAirplay } from 'react-icons/fi';

import Button from '../../../components/common/Button';

describe(`${Button.name}`, () => {
  describe('tests', () => {
    let wrapper, onClickSpy;

    beforeAll(() => {
      jest.spyOn(reactRedux, 'useSelector').mockImplementation(() => false);

      onClickSpy = jest.fn();

      wrapper = shallow(<Button text="A text" onClick={onClickSpy} />);
    });

    it('should render', () => {
      expect(wrapper.isEmptyRender()).toBeFalsy();
      expect(wrapper.find('.button')).toHaveLength(1);
    });

    it('should call onClick when clicked', () => {
      wrapper.find('.button').simulate('click');
      expect(onClickSpy).toBeCalledTimes(1);
    });

    it('should not call onClick when it is loading', () => {
      wrapper.setProps({ loading: true });
      wrapper.find('.button').simulate('click');
      expect(onClickSpy).toBeCalledTimes(0);
    });

    afterAll(() => {
      wrapper.unmount();
    });
  });

  describe('snapshots', () => {
    beforeEach(() => {
      jest.spyOn(reactRedux, 'useSelector').mockImplementation(() => false);
    });

    it('should render with text', () => {
      const tree = renderer
        .create(<Button text="A text" />)
        .toJSON();

      expect(tree).toMatchSnapshot();
    });

    it('should render borderless', () => {
      const tree = renderer
        .create(<Button text="A text" borderless />)
        .toJSON();

      expect(tree).toMatchSnapshot();
    });

    it('should render disabled', () => {
      const tree = renderer
        .create(<Button text="A text" disabled />)
        .toJSON();

      expect(tree).toMatchSnapshot();
    });

    it('should render loading', () => {
      const tree = renderer
        .create(<Button text="A text" loading />)
        .toJSON();

      expect(tree).toMatchSnapshot();
    });

    it('should render with icon', () => {
      const tree = renderer
        .create(<Button icon={<FiAirplay />} tooltip="A tooltip" />)
        .toJSON();

      expect(tree).toMatchSnapshot();
    });

    it('should render with marker', () => {
      const tree = renderer
        .create(<Button icon={<FiAirplay />} marker={<FiAirplay />} tooltip="A tooltip" />)
        .toJSON();

      expect(tree).toMatchSnapshot();
    });

    it('should render with marker in compact layout', () => {
      jest.spyOn(reactRedux, 'useSelector').mockImplementation(() => true);

      const tree = renderer
        .create(<Button icon={<FiAirplay />} marker={<FiAirplay />} tooltip="A tooltip" />)
        .toJSON();

      expect(tree).toMatchSnapshot();
    });
  });
});
