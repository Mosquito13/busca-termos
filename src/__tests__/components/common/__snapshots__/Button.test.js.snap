// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button snapshots should render borderless 1`] = `
<button
  className="button button--borderless"
  onClick={[Function]}
>
  <div
    className="button__content"
  >
    A text
  </div>
</button>
`;

exports[`Button snapshots should render disabled 1`] = `
<button
  className="button button--disabled"
  disabled={true}
  onClick={[Function]}
>
  <div
    className="button__content"
  >
    A text
  </div>
</button>
`;

exports[`Button snapshots should render loading 1`] = `
<button
  className="button button--loading"
  onClick={[Function]}
>
  <div
    className="button__content"
  >
    <span
      className="css-19vnu2x"
    />
  </div>
</button>
`;

exports[`Button snapshots should render with icon 1`] = `
<button
  className="button button--with-icon"
  onClick={[Function]}
  title="A tooltip"
>
  <div
    className="button__content"
  >
    <svg
      fill="none"
      height="1em"
      stroke="currentColor"
      strokeLinecap="round"
      strokeLinejoin="round"
      strokeWidth="2"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M5 17H4a2 2 0 0 1-2-2V5a2 2 0 0 1 2-2h16a2 2 0 0 1 2 2v10a2 2 0 0 1-2 2h-1"
      />
      <polygon
        points="12 15 17 21 7 21 12 15"
      />
    </svg>
  </div>
</button>
`;

exports[`Button snapshots should render with marker 1`] = `
<button
  className="button button--with-icon"
  onClick={[Function]}
  title="A tooltip"
>
  <div
    className="button__content"
  >
    <svg
      fill="none"
      height="1em"
      stroke="currentColor"
      strokeLinecap="round"
      strokeLinejoin="round"
      strokeWidth="2"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M5 17H4a2 2 0 0 1-2-2V5a2 2 0 0 1 2-2h16a2 2 0 0 1 2 2v10a2 2 0 0 1-2 2h-1"
      />
      <polygon
        points="12 15 17 21 7 21 12 15"
      />
    </svg>
  </div>
  <div
    className="button__marker"
  >
    <div
      className="icon"
    >
      <svg
        fill="none"
        height="1em"
        stroke="currentColor"
        strokeLinecap="round"
        strokeLinejoin="round"
        strokeWidth="2"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M5 17H4a2 2 0 0 1-2-2V5a2 2 0 0 1 2-2h16a2 2 0 0 1 2 2v10a2 2 0 0 1-2 2h-1"
        />
        <polygon
          points="12 15 17 21 7 21 12 15"
        />
      </svg>
    </div>
  </div>
</button>
`;

exports[`Button snapshots should render with marker in compact layout 1`] = `
<button
  className="button button--with-icon"
  onClick={[Function]}
  title="A tooltip"
>
  <div
    className="button__content"
  >
    <svg
      fill="none"
      height="1em"
      stroke="currentColor"
      strokeLinecap="round"
      strokeLinejoin="round"
      strokeWidth="2"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M5 17H4a2 2 0 0 1-2-2V5a2 2 0 0 1 2-2h16a2 2 0 0 1 2 2v10a2 2 0 0 1-2 2h-1"
      />
      <polygon
        points="12 15 17 21 7 21 12 15"
      />
    </svg>
  </div>
  <div
    className="button__marker"
  >
    <div
      className="icon"
    />
  </div>
</button>
`;

exports[`Button snapshots should render with text 1`] = `
<button
  className="button"
  onClick={[Function]}
>
  <div
    className="button__content"
  >
    A text
  </div>
</button>
`;
